### PR TITLE
REF: prepare (upcast) scalar before dispatching to arithmetic array ops

### PR DESCRIPTION
**Pull request by jorisvandenbossche**
This moves the potential upcasting of python/numpy scalars to pandas scalars out of the actual array op  (`arithmetic_op`), but moves it up to the Series / DataFrame level before calling the array op. This avoids calling it repeatedly on the same scalar for each column.

### CODE REVIEW

**Comment by jbrockmendel** [comment]
How costly is this?  I get that its a valid optimization, but it breaks the abstraction of the self-contained op.

**Comment by jorisvandenbossche** [comment]
> How costly is this? 

In the same `FrameWithFrameWide` benchmark (as mentioned in https://github.com/pandas-dev/pandas/pull/39820#issuecomment-825185163), calls to `_maybe_upcast_for_op` currently take around 6% on master. But I expect this number to become (relatively) bigger with other optimizations merged (like #39820). For reference, the actual `add` operator only takes around 18% in this benchmark, with everything else being overhead that can be partly optimized away. 

> I get that its a valid optimization, but it breaks the abstraction of the self-contained op.

We already do other preparation steps before the actual array_op as well (extract_array, extracting zerodim, checking length, wrapping datetimelike, etc). So I don't think the abstraction holds right now, and there is certainly an opportunity to organize this better. For example, we can ensure to have a clear set of steps (factored out as consistent helper functions) that need to be taken in the different code paths where `array_op` gets called (eg both in series op as frame op)

File: pandas/core/ops/array_ops.py [hunk 1]
```diff
@@ -422,7 +422,7 @@ def get_array_op(op):
         raise NotImplementedError(op_name)
 
 
-def _maybe_upcast_for_op(obj, shape: Shape):
+def prepare_scalar_for_op(obj, shape: Shape):
```
Comment (jbrockmendel) [inline pandas/core/ops/array_ops.py:425]: i think to be pandas-idiomatic might need to keep a "maybe"?

**Comment by jbrockmendel** [comment]
> We already do other preparation steps before the actual array_op 

Fair enough.

Generally LGTM.  Should this wait until after the ensure_wrapped_if_datetimelike PR so that can be folded into this helper?

File: pandas/core/ops/__init__.py [hunk 2]
```diff
@@ -428,6 +429,7 @@ def f(self, other, axis=default_axis, level=None, fill_value=None):
 
         axis = self._get_axis_number(axis) if axis is not None else 1
 
+        other = maybe_prepare_scalar_for_op(other, (self.shape[axis],))
```
Comment (jbrockmendel) [inline pandas/core/ops/__init__.py:432]: why self.shape[axis] instead of self.shape?

Comment (jorisvandenbossche) [reply pandas/core/ops/__init__.py:432]: Because `self` is a dataframe, and we want a 1D shape (rows or columns depending on the `axis`)

**Comment by jreback** [comment]
looks fine. this helps with perf?

File: pandas/core/ops/__init__.py (hunk 2 above)
Comment (jbrockmendel) [reply pandas/core/ops/__init__.py:432]: passing shape only matters when we have a scalar, in which case broadcasting to self.shape is simpler

**Comment by jorisvandenbossche** [comment]
> this helps with perf?

Yes, see https://github.com/pandas-dev/pandas/pull/40479#issuecomment-825191578

File: pandas/core/ops/__init__.py (hunk 2 above)
Comment (jorisvandenbossche) [reply pandas/core/ops/__init__.py]: OK, changed to self.shape

**Comment by jreback** [comment]
thanks @jorisvandenbossche

**Pull request merged**